libraries/AP_HAL_Linux/UARTDriver.cpp

```diff
@@ -417,7 +334,7 @@ int UARTDriver::_write_fd(const uint8_t *buf, uint16_t n)
     ret = _device->write(buf, n);
 
     if (ret > 0) {
-        BUF_ADVANCEHEAD(_writebuf, ret);
+        _writebuf.advance(ret);
```

Comment:
That would be a mistake because it would allow the ringbuffer to overwrite this buffer segment prior it's consumption by this function, which only happens the line above the if on this function. 

_write_fd() only exists to share code between two different cases of the caller function, that's why to have advance() called here isn't an real issue, it's private.

Which doesn't means that I like the overall implementation, but the real issue is not related with advance() and more to the fact that the return of this function isn't even checked, but again, the intention is to avoid change behaviour on this patch set. We can address this once this patch set is merged.